README.md
# CIHAutoNox

Tested on Kodi 16.2

This is a modified version of scope nox by funkd.. which is a modified version of AeonNox by BigNoid.
Features:

    - Can tag files with correct aspect ratio and audio codec: Atmos, DTS-X. in the .nfo file, add a genre containing:
        atmos, dts-x, AR133, AR137, AR166, AR178, AR185, AR220, AR235, AR240, AR255, AR276 (** NOT CASE SENSITIVE **)
        i.e. <genre>Dolby Atmos</genre>, or <genre>atmos AR235</genre>, or <genre>Atmos DTS-X</genre><genre>AR235</genre>
    - Mask (black out) areas outside of the specified aspect ratio.
    - Place the OSD and the Seekbar at the top and bottom of the screen
    - Automatically detect hardcoded black bars and calculate the aspect ratio of the playing video.
    - Can Manually specify the aspect ratio of the playing video.
    - Aspect ratio supported: 2.35, 2.0, 1.85, 16:9, 1.33. 
    **** Requires Kodi addon KodiCallback for autoaspect ratio detection ***
    - Can be used to control/recall projector settings specific to the AR of the playing video.

Idealy used for CIH zoom setups. 

Using the normal Kodi skins in this cases causes part of the skin to spill outside the scope frame onto walls.

This version of the skin contains all the GUI within an 800/820 pixel height. 


**************************
** Installation and Use **
**************************

1) Download the zip files to the Kodi device.<br>
    a) script.cihautonox.tools.zip<br>
    b) skin.cihautonox.1.0.zip <br>
<br>
2) Goto Addons, "Install from Zip" and install the two zip files<br>
<br>
3) Install the Kodi Callbacks addon. http://kodi.wiki/view/Add-on:Kodi_Callbacks<br>
    Available through the official repo (isengard forward) as of 22-04-2016<br>
<br>
4) Configuring KodiCallBacks:<br>
    Task 1: builtin - RunScript(script.cihautonox.tools,autoaspect)<br>
    Task 2: buitin - RunScript(script.cihautonox.tools,133stop)<br>
    Event 1: On PlaybackStarted - Task1<br>
    Event 2: On PlaybackStopped - Task2 (so we can remove the 1.33 side masks for the gui)<br>
<br>
5) Play a file and (hopefully) watch the magic happens.

***********************************************************************
** changing the projector zoom settings with the aspect ratio change **
***********************************************************************
The cool thing would be to control the projector zoom setting based on the video file.
the JVC-RS400 projector has lens memory settings. I just set up the memory slots for
2.35, 2.0, 1.85, 16:9, 1.33. 
These lens memory settings can be recalled via ip control. Unfortunately, IP control 
for the RS400 can only take request from one source at a time. i.e., kodi, and another
remote control app cannot both control the RS400.

The solution is to use a gateway. I just use the simplecontrol app. I have activities set
up to recall the various zoom settings. (Named Mask235, Mask200, Mask185, Mask169 in my case.)
Then on changing the aspect ratio on the skin, the script also tells simple control to run
the associated activity.

********************************
** Changing the Masking sizes **
********************************
custom_1131_ScopeBars133.xml - contains the mask for 133 aspect ratio<br>
custom_1132_ScopeBars235.xml - contains the mask for 235 aspect ratio<br>
custom_1133_ScopeBars200.xml - contains the mask for 200 aspect ratio<br>
custom_1134_ScopeBars185.xml - contains the mask for 185 aspect ratio.<br>
16:9 obviously does not need masking as it is the whole panel size.<br>

*************************
** Other files of note **
*************************
VideoOSD.xml - contains the OSD and aspect ratio changing buttons.
DiaglogSeekBar.xml - contains the code for the seek bar in the bottom of the screen.




skin.cihautonox.1.0-older/changelog.txt
[B]5.6.0[/]
- Fixed a problem with black overlay texture.

[B]5.5.9[/B]
- Updated language files from Transifex

[B]5.5.8[/B]
- Added support for Extended Info script - thx Steveb!
- Add BigList view - thx Steveb!
- Add support for Hebrew font - thx filthycatbox!
- Fix empty submenu for Custom menu item #5

[B]5.5.7[/B]
- Video Info Dialog: Converted actor list to thumb panel.
- Info panel: fix navigation to textbox/media menu on music views.
- Added viewtype "Wall" - thx phil65!
- Fixed: Text markup tags showing in some addons.

[B]5.5.6[/B]
- Some alignement fixes.
- Removed XBMC references.
- Updated language files from Transifex

[B]5.5.5[/B]
- Compatible with latest nightlies.
- Fixed alignment of tvshow unwatched episode count.

[B]5.5.4[]/B]
- Compatible with latest nightlies.

[B]5.5.3[/B]
- Compatible with latest nightlies.

[B]5.5.2[/B]
- Support international keyboard layouts in the virtual keyboard.
- Compatible with latest nightlies.

[B]5.5.1[/B]
- Add "Profiles" to the settings window.
- Compatible with latest nightlies.

[B]5.5.0[/B]
- Add Helix skinning engine changes.

[B]5.1.6[/B]
- List view for Songs and Albums: diffuse starrating with selected color when listitem is selected or playing.
- Landscape view: Fix invisible scrollbar + fix nav to plot scrollbar.
- Video information: Restore toggle on focused plot (full plot/outline).
- InfoWall: fix unreachable media menu in certain conditions.
- Shutdownmenu: Close all dialogs before going to settings.

[B]5.1.5[/B]
- Revert movement of fixed home list.
- Fix unreadable filestacking dialog.

[B]5.1.4[/B]
- The 3D button now opens a popup for selecting 3D modes instead of just an on/off toggle.
- Hide the paused icon on OSD when subtitle search is active.
- Add support for other US rating scrapers then Universal Movie Scraper.
- Add support for Australian mpaa ratings - thx kidhazy.
- Show landscape art whenever available in landscape view.
- Removed background setting in videos window.
- Removed Extra-Fanart support.
- Changed: Background will show fanart when available throughout the skin and fallback to the images setup in the general skin settings.
- Fixed default fanart showing when dialogs were opened on the Home window.
- Added setting to disable fanart troughout the skin: found in General-->Setup background images.
- Added scrollbar to ShowCase/Landscape infopanel plot.
- Several usability fixes.

[B]5.1.3[/B]
- Fixed empty fanart on 15 days layout (Next Aired addon).
- Smaller font for secondary label on List for non-media content.
- Fixed various "XFILE::CDirectory::GetDirectory - Error getting" log spam.
- Add option to use video playlists as backgrounds on the Home window. 
- Fix cyrillic uppercase. - thx flaterichd.
- Fix for at least three subtitle services -thx highlandr.
- Updated language files from Transifex.
[B]Attention: If you have Live Cases set as background, please choose them again in the background customizer for the home menu item.[/B]

[B]5.1.2[/B]
- Default fanart no longer is visible when dialogs are active.
- Add musicvideo support for video extras and tvtunes addon. (Note that there's (still) no support for hiding the videoextras button if no extras are available)
- Fix navigation in Weather window.
- Tweak music tripanel layout.
- Add lowlist/icons/tripanel to music-files section.

[B]5.1.1[/B]
- Added Red theme - thx bryanbrazil.
- Fix invisible floortexture in picture icon view.
- List view on episode level now shows stream info and plot.

[B]5.1.0[/B]
- Added weather and system infoline option.
- Usability tweaks for the manage submenu dialog (script.skinshortcuts).
- Adding a default select action for the main menu now uses script.skinshortcuts - thx Unfledged.
- Fixed InfoPanel button in ShowCase artists view.
- Fixed missing watched indicator when Slim List is selected.
- PVR, recordings view: Add plot and fix watched indicator.
- Add Live DVD Cases for random albums and songs.
- BannerWall now has a 3x7 grid layout.
- Add Episode view for episodes  - thx bryanbrazil.
- Add setting to customize the backgrounds images shown for all sections, if there's no fanart available. (Skin settings-->General-->Setup background images).
  This replaces the fallbacks to whatever was set on the Home menu; when a menu item was re-ordered it would show the fallbacks from the old menu item.
- Fix time seek display on screen.
- Updated language files from Transifex

[B]5.0.4[/B]
- Make changes for skin support next aired script.
- Remove custom submenu option. Skin shortcuts script is fast enough on every system.
- Remove redundant channelname pop-up from live tv.
- Fix artist info feed texture in visualisation when main menu is toggled in lower position.
- Fixed grapphical error in context menu if there was only one item.
- The weather widget is no longer on top when the main menu is in lower position.
- Fix case-sensitive error custom submenus on *nix systems.
- Fix fanart on music viz with waveform, opengl spectrum and directx spectrum as visualization.

[B]5.0.3[/B]
- Updated language files from Transifex

[B]5.0.2[/B]
- Add episode count to TriPanel/List season view.
- Corrected pointer texture.
- Unify font size in dialogs.
- Orange theme: yellow selected color instead of bright orange.
- PVR: Add popup when changing channels.
- PVR: Add missing textures channel manager.
- PVR: Fix missing texture on group change dialog.
- Hide HD/SD/3D tag in title in views that have a resolution flag.

[B]5.0.1[/B]
- Added unwatched episodes count indicator.
- Added HD/SD/3D indicator to floorlabel.
- Brighter textcolor for widget text.
- Logo view wrap/fixed setting.
- Moved to different way of handling custom submenu's. If already added custom submenu's have disappeared, make a change in one of the submenus, save it and all previous custom items will re-appear.

script.cihautonox.tools/addon.py
import json
import urllib
import urllib2
import sys
import xbmc
import xbmcaddon
import xbmcgui
import os.path

addon       = xbmcaddon.Addon()
addonname   = addon.getAddonInfo('name')
capture     = xbmc.RenderCapture()
myplayer    = xbmc.Player()

# interface to run activity for simple control
SCCommand = 'http://<your simplecontrol hub ip address>:47147/api/v1/runactivity'
SCActivity_widescreen = "<your simple control activity UUID>"
SCActivity_20screen   = "<your simple control activity UUID>"
SCActivity_185screen  = "<your simple control activity UUID>"
SCActivity_169screen  = "<your simple control activity UUID>"
SCActivity_133screen  = "<your simple control activity UUID>"
SCActivity_lowlights  = "<your simple control activity UUID>"
SCActivity_movielights = "<your simple control activity UUID>"

# capture Width & Height
# each captured pixel is 40wx20h of the screen pixel
# 1920 x 1080 = 1.77
# 1920 x 1038 = 1.85 (20) (line 0 dark)
# 1920 x 960 = 2.0 (60) (line 0 - 2 dark)
# 1920 x 817 = 2.35 (130) (line 0 - 5 dark)
CaptureWidth = 48
CaptureHeight = 54


# subtitle vertical positions for aspect ratio
Sub235 = 860
Sub200 = 1000
Sub185 = 1040
Sub169 = 1080
Sub133 = 1080



###########
#
# SCRunActivity
#   sends a json rpc to simple control to change projector zoom / lenshift /masking
#   In simplecontrol, set an activity to recall the projector lens memory
#   get the activity uuid by going to:
#   http://<simple control hub ip address>:47147/api/v1/activities
#   then copy the UUID to the above variable. i.e. SCActivity_169 screen
#
###########
def SCRunActivity (activity):
    return
    temp = '{"activity_uuid" : "' + activity + '"}'
    urllib.urlopen (SCCommand, temp)

###########
#
# PARdecrease
# PARincrease
# PARReset
#    needs custom Kodi. These ActionIDs are not in the standard Kodi yet.
#    will not break the script if it's not used.
#
###########
PARSteps = 11
def PARdecrease ():
    for i in range (1, PARSteps):
        xbmc.executebuiltin ('XBMC.Action(DecreasePARSilent)')

def PARincrease ():
    xbmc.executebuiltin('Skin.setbool (osdstretched)')
    for i in range (1, PARSteps):
        xbmc.executebuiltin ('XBMC.Action(IncreasePARSilent)')

def PARReset ():
    xbmc.executebuiltin ('XBMC.Action(ResetPAR)')
    xbmc.executebuiltin('Skin.reset (osdstretched)')

###########
#
# SubtitleVshiftSet
#   Set absolute subtitle conditions
#   works best with subtitle align manual
#
# does not work... likely problem with how kodi renders subtitles
###########

def SubtitleVshiftSet (_count):
    return
    xbmc.executebuiltin ('XBMC.Action(ResetSubtitleVshift)')
    for i in range (1, _count):
        xbmc.executebuiltin ('XBMC.Action(IncreaseSubtitleVshift)')

###########
#
# ARChanges routine
#   notifies the skin CIHAutoNox of the aspect ratio change
###########
def ARClearBooleans ():
    xbmc.executebuiltin('Skin.reset (osd133)')
    xbmc.executebuiltin('Skin.reset (osd169)')
    xbmc.executebuiltin('Skin.reset (osd185)')
    xbmc.executebuiltin('Skin.reset (osd200)')
    xbmc.executebuiltin('Skin.reset (osd235)')

def ARChange133():
    SCRunActivity (SCActivity_133screen)
    SubtitleVshiftSet (Sub133)
    ARClearBooleans()
    xbmc.executebuiltin('Skin.setbool (osd133)')

def ARChange169():
    SCRunActivity (SCActivity_169screen)
    SubtitleVshiftSet (Sub169)
    ARClearBooleans ()
    xbmc.executebuiltin('Skin.setbool (osd169)')

def ARChange185():
    SCRunActivity (SCActivity_185screen)
    SubtitleVshiftSet (Sub185)
    ARClearBooleans()
    xbmc.executebuiltin('Skin.setbool (osd185)')

def ARChange200():
    SCRunActivity (SCActivity_20screen)
    SubtitleVshiftSet (Sub200)
    ARClearBooleans()
    xbmc.executebuiltin('Skin.setbool (osd200)')


def ARChange235():
    SCRunActivity (SCActivity_widescreen)
    SubtitleVshiftSet (Sub235)
    ARClearBooleans ()
    xbmc.executebuiltin('Skin.setbool (osd235)')

##############
#
# CaptureFrame
#
# Capture a frame into variable "capture"
# global variable - capture, CaptureWidth, CaptureHeight
# returns - image byte array
#############
def CaptureFrame ():
    capture.capture(CaptureWidth,CaptureHeight, xbmc.CAPTURE_FLAG_CONTINUOUS)
        
    while (capture.getCaptureState() != xbmc.CAPTURE_STATE_DONE):
        capture.waitForCaptureStateChangeEvent ()
        
        if (capture.getCaptureState() == xbmc.CAPTURE_STATE_DONE):
            capturedImage = capture.getImage()
            return capturedImage

def CaptureFrame_new ():
    capture.Capture (CaptureWidth, CaptureHeight)
    capturedImage = capture.getImage (1000)
    return capturedImage

##############
#
# LineColorLessThan
#    _bArray: byte Array that contains the data we want to test
#    _lineStart: where to start testing
#    _lineCount: how many lines to test
#    _threshold: value to determine testing
# returns: True False
###############
def LineColorLessThan (_bArray, _lineStart, _lineCount, _threshold):
    __sliceStart = _lineStart * CaptureWidth * 4
    __sliceEnd = (_lineStart + _lineCount) * CaptureWidth * 4
    
    # zero out the alpha channel
    i = __sliceStart + 3
    while (i < __sliceEnd):
        _bArray[i] &= 0x00
        i += 4
    
    __imageLine = _bArray[__sliceStart:__sliceEnd]
    __result = all([v < _threshold for v in __imageLine])
    return __result

###############
#
# GetAspectRatioFromFrame
#   - returns Aspect ratio * 100 (i.e. 2.35 = 235)
#   Detects hardcoded black bars
###############
def GetAspectRatioFromFrame ():
    __aspectratio = int( (capture.getAspectRatio() + 0.005) * 100 )
    __threshold = 25
    
    # Analyze the frame only if the ratio is 16:9. 2.35 ratio files
    # would not have black bars hardcoded.
    if 140 < __aspectratio < 180 :
        # screen capture and test for an image that is not dark in the 2.40
        # aspect ratio area. keep on capturing images until captured image
        # is not dark
        while (True):
            __myimage = CaptureFrame()
            __middleScreenDark = LineColorLessThan (__myimage, 7, 2, __threshold)
            if __middleScreenDark == False:
                xbmc.sleep (200)
                break
            else:
                xbmc.sleep (100)
    
        # Capture another frame. after we have waited for transitions
        __myimage = CaptureFrame()
        __ar185 = LineColorLessThan (__myimage, 0, 1, __threshold)
        __ar200 = LineColorLessThan (__myimage, 1, 3, __threshold)
        __ar235 = LineColorLessThan (__myimage, 1, 5, __threshold)
        
        if (__ar235 == True):
            __aspectratio = 235

elif (__ar200 == True):
    __aspectratio = 200
        
        elif (__ar185 == True):
            __aspectratio = 185

return __aspectratio

###############
#
# AutoaspectChange
#   must be called by kodi on playback start
#   the easiest way is to use the kodi addon: KodiCallback and set Task 1 to:
#   task: builtin
#   kodi builtin function: RunScript(script.cihautonox.tools,autoaspect)
#   then set Event 1: on Playback started
#            task for Event 1: Task 1
#
###############
def AutoaspectChange ():

    aspectratio = GetAspectRatioFromFrame ()
    PARReset()
    
    # 16:9 screen
    if (150 <= aspectratio <= 180):
        ARChange169()
    
    # 1.33 screen
    elif (130 <= aspectratio <= 140):
        ARChange133()
    
    # 1.85 screen
    elif (181 <= aspectratio <= 189):
        ARChange185()
    
    # 2.0 Screen
    elif (190 <= aspectratio <= 210):
        ARChange200()
    
    # 2.35 Screen
    elif (aspectratio > 210):
        ARChange235()

###############
#
# AutoStretch
#   stretch 1:33 -> no stretch
#   stretch      -> 2.0
#   No stretch 2.35 and up
#   same use as above. Could set this as task 2
#
#   *** Does not work in stock Kodi. I am awaiting my pull request to be merged into kodi ***
###############
def AutoStretch ():
    
    aspectratio = GetAspectRatioFromFrame ()
    PARReset()
    
    # 1.33 screen no stretch
    if (130 <= aspectratio <= 140):
        ARChange133()
    
    #stretch 16:9, 1.85:1 to 2.0
    elif ( 150 <= aspectratio <= 189):
        ARChange200 ()
        PARincrease()
    
    #stretch 2.0 to 2.35
    elif ( 190 <= aspectratio <= 210):
        ARChange235 ()
        PARincrease ()

    # 2.35 screen no stretch
    elif (aspectratio >210):
        ARChange235 ()


######################################
######################################

# Set AspectRatio masking on
xbmc.executebuiltin('skin.reset (scopemask)')
xbmc.executebuiltin('Skin.setbool (ARmask)')

#####################
# Testing
#####################

if sys.argv[1] == 'test':
    
    aspectratio = GetAspectRatioFromFrame ()
    aspectratio2 = int ((capture.getAspectRatio () + 0.005) * 100)

    _info = xbmc.getInfoLabel('Player.Process(VideoDAR)')
    _info2 = xbmc.getInfoLabel('Player.Process(videoheight)')
    WIN = xbmcgui.Window(10000)
    line1 = 'Calculated Aspect Ratio = ' + str(aspectratio)
    line2 = 'Player Aspect Ratio = ' + str(aspectratio2)
    line3 = ' '
    xbmcgui.Dialog().ok(addonname,line1, line2, line3)

###################################
# end Testing
###################################


#----------------
# manual scope / masking buttons
#---------------
if sys.argv[1] == 'widescreen':
    ARChange235()

if sys.argv[1] == 'mediumscreen':
    ARChange200()

if sys.argv[1] == '185screen':
    ARChange185()

if sys.argv[1] == 'standardscreen':
    ARChange169()


#----------------
# autoaspect ratio change
#-----------------

if sys.argv[1] == 'autoaspect':
    AutoaspectChange()


if sys.argv[1] == 'autostretch':
    AutoStretch()


#----------------
# shrink
#   unstretch video if it has been stretched.
#-----------------
if sys.argv[1] == 'shrink':
    AutoaspectChange()

#----------------
# stretch
#-----------------
if sys.argv[1] == 'stretch':
    aspectratio = GetAspectRatioFromFrame ()
    

    PARReset()
    
    #stretch 133 to 16:9
    if (130 <= aspectratio <= 140):
        ARChange169()
        PARincrease()
    
    #stretch 16:9, 1.85:1 to 2.0
    elif ( 150 <= aspectratio <= 189):
        ARChange200 ()
        PARincrease()

    #stretch 2.0 to 2.35
    elif ( 190 <= aspectratio <= 210):
        ARChange235 ()
        PARincrease ()

#---------------
# AR133Stops
#   if playing aspectratio 133, and stops, opens mask up to 169
#   if this is not done, the aspect ratio will be masked at 1.33
#---------------
if sys.argv[1] == '133Stop':
    aspectratio = GetAspectRatioFromFrame ()
    if (130 <= aspectratio <= 140):
        ARChange169()

